src/ssl.c

```diff
@@ -15479,34 +15593,39 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
 
 #ifdef SESSION_CACHE_DYNAMIC_MEM
     cacheSession = sessRow->Sessions[idx];
-    if (cacheSession) {
-        XFREE(cacheSession, sessRow->heap, DYNAMIC_TYPE_SESSION);
-        cacheSession = NULL;
-    }
-    cacheSession = (WOLFSSL_SESSION*) XMALLOC(sizeof(WOLFSSL_SESSION), sessRow->heap,
-                                              DYNAMIC_TYPE_SESSION);
     if (cacheSession == NULL) {
-        return MEMORY_E;
+        cacheSession = (WOLFSSL_SESSION*) XMALLOC(sizeof(WOLFSSL_SESSION),
+                                         sessRow->heap, DYNAMIC_TYPE_SESSION);
+        if (cacheSession == NULL) {
+            return MEMORY_E;
+        }
+        XMEMSET(cacheSession, 0, sizeof(WOLFSSL_SESSION));
+        sessRow->Sessions[idx] = cacheSession;
     }
-    XMEMSET(cacheSession, 0, sizeof(WOLFSSL_SESSION));
-    sessRow->Sessions[idx] = cacheSession;
 #else
     cacheSession = &sessRow->Sessions[idx];
 #endif
 
 #ifdef HAVE_EX_DATA
-    if (cacheSession->ownExData) {
+    if (overwrite) {
+        /* Figure out who owns the ex_data */
+        if (cacheSession->ownExData) {
+            /* Prioritize cacheSession copy */
+            XMEMCPY(&addSession->ex_data, &cacheSession->ex_data,
```

Comment:
`overwrite == 1` means that the session already exists in cache. The `ex_data` in cache is the single source of the real `ex_data`. It can be shared between objects.

Reply:
I understand. Thanks.